### ISSUE TITLE
Example for getJoinData

### ISSUE THREAD

**Issue by prakash-indorkar** [jdgamble555/adv-firestore-functions#13]
Hi, I understood the use of createJoinData and updateJoinData ( very straight forward ). 
However, I'm looking for usage of the getJoinData function.

use case: ( a chat module )

users => { email, displayName, photoURL }
conversations => { lastMessage, userIds, userInfo: { 'userId': { displayName, photoURL} }, dateTime}
messages => { text, dateTime, convoId }

In the conversations table, I'm storing an array of userIds ( user involved in the conversation ), I want to get their displayName and photoURL in userInfo map. So if two users are there, both their details will be stored in userInfo map having key as their userId. From the frontend I can do that easily, but I want to use getJoinData to get data for both users and store it in the userInfo map.

I tried something like this..

```javascript

exports.createUserInfoInConversation = functions.firestore.document('/conversations/{docId}')
    .onWrite(async (change, context) => {
           if (isTriggerFunction(change, context)) {
                return null;
            }
           const joinFields = ['displayName', 'photoURL'];
           const userIds = getValue(change, 'userIds');
           const data = new Map();
           userIds.forEach(userId => {
              const userRef = db.collection(`users/${userId}`);
              data['${userId}'] = await getJoinData(change, userRef, joinFields, 'userInfo');
           });
           await triggerFunction(change, data);
}
```

later I want to write an update function as well to update if any user data changes, but I couldn't get this working.

Please help me with your expertise, Thank you in advance, you have already made things easy for me. I'm using your package in full and also telling my other friends to see usability for them.

**Comment by jdgamble555** [comment]
This is not an issue, but help with programming.  Please limit your questions to basic understanding of functions or problems.  I frankly don't have time. 

You said you couldn't get this working.  What is the problem?  What are the errors? etc...

You could easily debug this by:

```typescript
const test = await getJoinData(change, userRef, joinFields, 'userInfo');
console.log(test);
```

Try something like this:

```typescript
const userIds = getValue(change, 'userIds');
const data = new Map();
const users = [];
userIds.forEach(userId => {
  const userRef = db.collection(`users/${userId}`);
  const user = await getJoinData(change, userRef, joinFields, 'userInfo');
  users.push(user);
});
data['users'] = users;
await triggerFunction(change, data);

**Comment by prakash-indorkar** [comment]
Hi, I've actually tried all the options. 

console.log(await getJoinData(change, userRef, joinFields, 'userInfo'));  ===-> prints->  **{ }**    empty map.

getJoinData does not return anything. It could be a bug!. (If I use createJoinData - it works as expected, but creates 1 user data)

data['users'] = users;   can not be used to save using triggerFunction ( its giving error: Value for argument "data" is not a valid Firestore document.)

 that's the reason I was asking for help.
 
 No worries. I'll have to figure it out on my own. Thanks for your help. Regards.

**Comment by jdgamble555** [comment]
You can see [here](https://github.com/jdgamble555/adv-firestore-functions/blob/master/src/shared/joins.ts) that **createJoinData** actually contains **getJoinData**, so **create** cannot work without **join** working, which means your query must be off.

Make sure to set **data** is set to an object, and not a **map**.  While it might work, I have not tested it that way.  

Please post exact code with **create** and with **join** so we can see if it is a bug or not.  If it is a bug, I am happy to look at it and fix it.  I need exact code and errors.

If you are not using the **array** code from above, it will probably only get 1 user data.

**Comment by prakash-indorkar** [comment]
Hi.. I have figured it out .. by executing multiple times with changes..  This is what works for me

```javascript
exports.createUserInfoInConversation = functions.firestore.document('/conversations/{docId}')
    .onWrite(async (change, context) => {
        if (isTriggerFunction(change, context)) {
            return null;
        }

        if (createDoc(change)) {
            const joinFields = ['displayName', 'photoURL'];
            const userIds = getValue(change, 'userIDs');
            const data = {};
            userIds.forEach(async (userId) => {
                const userRef = db.doc(`users/${userId}`);
                data['userInfo'] = await getJoinData(change, userRef, joinFields, userId);
                await triggerFunction(change, data);
            });
        }
        return null;
    });
```

what I have noticed is - if I print console.log(data) outside of forEach loop .. it prints empty {}. ( I tried with your code as well the array was also empty). That's why I thought it could be a bug. And when I try to use triggerFunction(change, data); outside as well, it again did not works. ( I tried it more than 100 times) finally, I thought I'll call for each user update and it worked.

Now as per your instructions, triggerFunction(change, data); should be called only once..  but it did not save data. With your code it was giving error 
```
Error: Value for argument "data" is not a valid Firestore document. Input is not a plain JavaScript object.
```
Now that it works.. all the functions execute multiple times until it finishes completely. You can suggest me better approach. 
Thanks for looking into this.. You can close this issue now, as this was not a bug. 
Now I'll have to figure out how to update from user change because it's again a nested map.

Regards, Prakash

**Comment by jdgamble555** [comment]
Yes, you should take **triggerFunction** out of the loop.  It should **ALWAYS** only run once, otherwise you are incurring multiple writes you do not need to incur.  Again, use my code and set the data to an array, and update the array.  It may also be that you should also use a regular foreach loop.  In this case, you do not want the **async** calls running at the same time, you want them to finish first, then run the triggerFunction.  See [here](https://stackoverflow.com/a/37576787/271450).

Try this:

```typescript
const userIds = getValue(change, 'userIds');
const data = {};
const users = [];
for (const userId of userIds) {
  const userRef = db.collection(`users/${userId}`);
  const user = await getJoinData(change, userRef, joinFields, 'userInfo');
  users.push(user);
}
data['users'] = users;
await triggerFunction(change, data);
```

You also don't need ```if (createDoc(change))``` as this is already in the code to prevent the call.

**Comment by prakash-indorkar** [comment]
Thanks, mate!...

changing to the normal for loop did the trick.

```javascript
exports.createUserInfoInConversation = functions.firestore.document('/conversations/{docId}')
    .onWrite(async (change, context) => {
        if (isTriggerFunction(change, context)) {
            return null;
        }

            const joinFields = ['displayName', 'photoURL'];
            const userIds = getValue(change, 'userIds');
            const data = {};
            const users = {};
            for (const userId of userIds) {
                const userRef = db.doc(`users/${userId}`);
                const user = await getJoinData(change, userRef, joinFields, userId);
                users[userId] = user[userId];
            }
            data['userInfo'] = users;
            await triggerFunction(change, data);
        return null;
    });
```

This way I have 'userInfo' field, userId as key and joinFields as values. (easy to use on the front end).

Thank you for your time and efforts.  ( I'm basically using dart and flutter for front-end, so it gets mixup sometimes).

If you do not mind, I need to write an update function as well. (**Please ignore if I'm taking too much of your time**)

Here's what I did for the update.. 

```javascript
exports.updateUserRefData = functions.firestore.document('/users/{docId}')
    .onWrite(async (change, context) => {
             if (isTriggerFunction(change, context)) {
                return null;
             }
             if (valueChange(change, 'displayName') || valueChange(change, 'photoURL')) {
                    const docId = context.params.docId;
                    const queryRef = db.collection('conversations').where('userIds', 'array-contains', docId);
                    const joinFields = ['displayName', 'photoURL'];
                    await updateJoinData(change, queryRef, joinFields, `userInfo.${docId}`);
             }
}
```
It creates a new field. 

What I feel .. updateJoinData(change, queryRef, joinFields, nested.field);  should be able to accept nested field, like  field.key or key[key]. I see it only takes field name as 'string' where it updates joinData. Then above code can update fields easily with 'userInfo[docId]'.  Like firebase accepts nested fields using .(dot) in querie's where condition.

**Comment by jdgamble555** [comment]
Agreed. Should now support dot notation. Let me know if it works now with 1.9.0.

**Comment by prakash-indorkar** [comment]
Thanks @jdgamble555 . It works like charm!. 

I have only one concern now.. _events collection still generating lot of docs .. (not an infinite loop though). I see it is mainly because of `await triggerFunction(change)`. I have a universal function, only for updating createdAt and updatedAt. It executes too frequently on every call.
```javascript
exports.autoCreatedAtUpdatedAt = functions.firestore.document('{colId}/{docId}')
    .onWrite(async (change, context) => {
        const colId = context.params.colId;

        if (colId.startsWith('_') || isTriggerFunction(change, context)) {
            return null;
        }
        await triggerFunction(change);
        return null;
    });
```

my index.js looks this 

```javascript
const autoCountCollection = require("./autoCountCollection");
const autoCreatedAtUpdatedAt = require("./autoCreatedAtUpdatedAt");
const autoUpdateRefUserData = require("./autoUpdateRefUserData");
const autoUpdateRefSharingData = require("./autoUpdateRefSharingData");
const sharingsCounter = require("./sharingsCounter");


//collection counter function
exports.autoCountUsers = autoCountCollection.autoCountUsers;
exports.autoCountSharings = autoCountCollection.autoCountSharings;


// Maintain active/consumed sharing counter under _counters/sharings
exports.activeSharingsCounter = sharingsCounter.activeSharingCounter;
exports.consumedSharingsCounter = sharingsCounter.consumedSharingCounter;
//update isActive depending on isConsumed or isDeleted or isExpired 
exports.updateActiveSharing = sharingsCounter.updateActiveSharing


// Update CreatedAt and UpdateAt dateTime - universal function
exports.autoUpdateDateTime = autoCreatedAtUpdatedAt.autoCreatedAtUpdatedAt;


// Get and update UserInfo in sharings
exports.createUserInfoInSharing = autoUpdateRefUserData.createUserInfoInSharing;
exports.updateUserInfoInSharing = autoUpdateRefUserData.updateUserInfoInSharing;


//Get and update UserInfo in conversations
exports.createUserInfoInConversation = autoUpdateRefUserData.createUserInfoInConversation;
exports.updateUserInfoInConversation = autoUpdateRefUserData.updateUserInfoInConversations;


// Get and update SharingInfo in new conversation
exports.createSharingInfoInConversation = autoUpdateRefSharingData.createSharingInfoInConversation;
exports.updateSharingInfoInConversation = autoUpdateRefSharingData.updateSharingInfoInConversation;

```

You see almost all functions execute all time and autoCreatedAtUpdatedAt executes along with them multiple times (they may not be updating on database, but _events collection surely increasing with new records every time (that's multiple db writes).

If I remove  **autoCreatedAtUpdatedAt** function, it reduces the number of calls significantly (and also fewer writes in _events). I have the option to update these dates from front-end anyway, so can still be a choice. 

Otherwise, I'm pretty happy with the outcome. Thank you once again for keeping up with me. I'll give a try on using other function too like the tags, categories, indexing. etc and will get back if required. Hope to see your blog with new examples.

Regards, Prakash

**Comment by jdgamble555** [comment]
First, this is a new issue.  Please do not keep adding new issues to this thread!

I am not sure you understand how the functions work.  Neither **triggerFunction** nor **isTriggerFunction** create any events.  You can verify that [here](https://github.com/jdgamble555/adv-firestore-functions/blob/master/src/shared/tools.ts) in my code. 

Only counter functions should create events. If **updatedAt** or **createdAt** gets triggered more than once, it does not matter, as the dates will still be accurate.  If a **counter** function gets triggered more than once, the counter will be off, hence the need for **event** management.

I have recently updated my advice on universal date functions [here](https://stackoverflow.com/a/60963531/271450).  It is always better to do this on the front end, but verifying it with **firestore rules**.  This prevents frontend hackers.  I no longer recommend my own function.

### However, your real problem has nothing to do with dates.

I would need to see your code.  I suspect you are running **eventExists** when you don't have to, and too many universal functions.  **YOU CANNOT KEEP ACURATE COUNTS WITHOUT EVENT MANAGEMENT**.  However, I suspect you could limit or get rid of **universal** functions all together.  Only **counter** functions or **heavy-write** functions should use events.  

It really depends on your use case, and you have to weigh the pros and cons of each function.  The goal is to limit the amount of times **eventExists** is ran, while also keeping counter functions in separate functions all together for accuracy.  

For counters, you can limit checking events by adding:

```typescript
if (shiftDoc(change) && await eventExists(context)) {
    return null;
}
```
You only need to run a **counter** function or **eventExists** when adding or deleting a record.  For **full text search**, however, you cannot do this with the **eventExists** function.

It is all about management of your code and is situation-dependent.  

Good luck!
